### DISCUSSION TITLE
Custom node view not rendering contents until you click on it

### DISCUSSION THREAD

**Discussion by chasinhues** [ueberdosis/tiptap#1675] [General Questions]
Hi all, 

I've run into an issue that I am really struggling with at the moment.

We're using the code block lowlight extension to get syntax highlighting on the code in the editor. It looks incredible, but there's this issue where the content of the code block isn't rendered until you click on the editor. The video below illustrates the behaviour:

https://www.loom.com/share/2e0cdd45b4de4b48b4a1449cfc8d36f5

Note that the console logs show that the code block extension is receiving the content.

I believe @WilliamIPark raised the same issue in Discord, so seems I'm not the only one: https://discord.com/channels/818568566479257641/818569721934774272/870627395328172043

The code block extension (simplified, still reproduces the issue):
```jsx
import { NodeViewWrapper, NodeViewContent, ReactNodeViewRenderer, NodeViewRendererProps } from "@tiptap/react";
import CodeBlockLowlight from "@tiptap/extension-code-block-lowlight";
import lowlight from "lowlight";

export const CodeblockComponent = (props: NodeViewRendererProps) => {
    return (
        <NodeViewWrapper>
            <NodeViewContent as="code" />
        </NodeViewWrapper>
    );
};

export const CodeblockExtension = CodeBlockLowlight.extend({
    addNodeView() {
        return ReactNodeViewRenderer(CodeblockComponent);
    },
}).configure({ lowlight });

```

Would really appreciate some help on this one - thanks!

**Comment by WilliamIPark** [comment]
Hey @chasinhues. Yeah, I run into this too and spent a fair bit of time trying to find a workaround for the issue.

I'll add in the clip I took myself of this issue to add to the discussion, which shows content in the data, but said content is not rendered to the UI until I click into the node.

https://user-images.githubusercontent.com/445688/128170455-484306e5-279c-4e38-b648-90fb02cc9163.mov

I suspect the cause is somewhat related to focus management. Block elements such as your CodeBlock look to be inserted fine, but inline or text elements won't be properly created if focus is dragged elsewhere in the document at the same time.

In my case, I believe the focus was being changed when I clicked on the blue area to create the new nodes. I was doing this with an onClick handler:

![Screenshot 2021-08-04 at 12 13 34 pm](https://user-images.githubusercontent.com/445688/128171504-4b29ab95-b79e-4c8f-be68-5ff105b0c6cc.png)

Which called this function:

![Screenshot 2021-08-04 at 12 14 38 pm](https://user-images.githubusercontent.com/445688/128171622-02494604-0b0b-47f9-952a-5fc2f69ffea7.png)

My fix, was to delay the creation of the node slightly, by wrapping it in a [requestAnimationFrame](https://developer.mozilla.org/en-US/docs/Web/API/window/requestAnimationFrame), but a setTimeout worked too.

![Screenshot 2021-08-04 at 12 16 13 pm](https://user-images.githubusercontent.com/445688/128171816-07a9b827-fa52-4c36-a8e4-4b881eacd3e5.png)

Hopefully this is of help. I'm glad you've created a minimal example of this (I was going to, when I got time), as this fix feels a bit hacky.

**Reply by chasinhues** [reply]
I wonder in your case if you need to `editor.chain().focus().insertContentAt(...).run()` instead of `editor.chain().insertContentAt(...).focus().run()`? (Note the different ordering)

I think calling `.focus()` before `.insertContentAt()` would focus the editor before trying to do any mutations to the DOM.

**Reply by WilliamIPark** [reply]
I've worded things poorly, sorry. 

The focus in the chain there isn't required at all (I tried both orders). That focus call is there because I do want the new block to be focused after creation in my use-case. 

The important part appears to be the delay to get this working. I've used the word 'focus', but it may be the ProseMirror 'selection' that is important. It's really hard to tell. Super hopeful that the main developers can shed some light on this.
